File: server/internal/schemas/sync.go
package schemas

import "time"

type SyncOpType string

const (
	SyncOpUpsert SyncOpType = "upsert"
	SyncOpDelete SyncOpType = "delete"
)

type SyncEntityType string

const (
	SyncEntityNote   SyncEntityType = "note"
	SyncEntityFolder SyncEntityType = "folder"
	SyncEntityTask   SyncEntityType = "task"
)

type SyncOperation struct {
	ID         string         `json:"id"` // Optional client op id for tracking
	Type       SyncOpType     `json:"type"`
	EntityType SyncEntityType `json:"entityType"`
	EntityID   string         `json:"entityId"`
	Payload    interface{}    `json:"payload,omitempty"`
}

type SyncRequest struct {
	LastSyncAt *time.Time      `json:"lastSyncAt,omitempty"`
	LastCursor *string         `json:"lastCursor,omitempty"`
	Ops        []SyncOperation `json:"ops"`
}

type IDMapping struct {
	LocalID  string `json:"localId"`
	ServerID string `json:"serverId"`
}

type SyncResponse struct {
	ServerTime     time.Time        `json:"serverTime"`
	NextCursor     string           `json:"nextCursor"`
	Notes          []NoteResponse   `json:"notes"`
	Folders        []FolderResponse `json:"folders"`
	Tasks          []TaskResponse   `json:"tasks"`
	Deleted        []SyncTombstone  `json:"deleted"`
	ProcessedOpIDs []string         `json:"processedOpIds"`
	IDMappings     []IDMapping      `json:"idMappings"`
}

type SyncTombstone struct {
	EntityType SyncEntityType `json:"entityType"`
	EntityID   string         `json:"entityId"`
}

File: desktop/src/types/sync.ts
import type { Folder } from "./folders";
import type { Note } from "./notes";

export type SyncOperation = {
  type: "upsert" | "delete";
  entityType: "note" | "folder";
  entityId: string;
  payload?: any;
};

export type SyncRequest = {
  lastSyncAt?: string;
  lastCursor?: string;
  deviceId?: string;
  ops: SyncOperation[];
};

export type SyncResponse = {
  serverTime: string;
  nextCursor: string;
  notes: Note[];
  folders: Folder[];
  deleted: { entityType: "note" | "folder"; entityId: string }[];
  processedOpIds: string[];
  idMappings: { localId: string; serverId: string }[];
};
